### PR TITLE
Finetuning GPT-2 with LoRA

### PR DESCRIPTION
**Pull request by abheesht17**
Resolves https://github.com/keras-team/keras-nlp/issues/1051

### CODE REVIEW

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 1]
```diff
@@ -0,0 +1,590 @@
...
+new LoRA layers.
+"""
+
+gpt2_backbone_layer_name = [
+    layer.name for layer in list(lora_model.layers) if "GPT2Backbone" in str(layer)
+][0]
+
+for layer_idx in range(lora_model.get_layer(gpt2_backbone_layer_name).num_layers):
+    # Change query dense layer.
+    original_query_layer = (
```
Comment (mattdangerw) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:415]: I think this whole block could be written a bit more readably. E.g.

```
decoder_layer = lora_model.backbone.get_layer(f"transformer_layer_{layer_idx}")
self_attention_layer = decoder_layer._self_attention_layer

self_attention_layer._query_dense = LoraLayer(...)

self_attention_layer._value_dense = LoraLayer(...)
```

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 2]
```diff
@@ -0,0 +1,590 @@
...
+    ) = LoraLayer(
+        original_value_layer,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
+        trainable=True,
+    )
+
+"""
+Let's now do a forward pass to make sure it is going through.
```
Comment (mattdangerw) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:453]: it is going through. -> we still have a valid chain of computation? or something like that

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+lora_model_memory_labels = gpu_memory_callback.labels
+
+"""
+And we are done fine-tuning the model! Before we generate text, let's compare
+the training time and memory usage of the two models. The training time of GPT-2
+on a 16 GB Tesla T4 (Colab) is 7 minutes, and for LoRA, it is 5 minutes, a 30%
+decrease. The memory usage of LoRA GPT-2 is roughly 3 times less than GPT-2.
+"""
+
+plt.plot(gpt2_lm_memory_labels, gpt2_lm_memory_usage, label="GPT-2")
```
Comment (mattdangerw) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:529]: awesome!

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 4]
```diff
@@ -0,0 +1,590 @@
...
+does not incur any additional inference latency. Let's understand why.
+
+Recall our LoRA equation: `output = W0x + b0 + BAx`. We can rewrite this as:
+`output = = Wx + b0 = (W0 + BA)x + b0`, where `W = W0 + BA`. This means that if
+we merge the weights of the original model and the adapter, we will be essentially
+doing the same computation as the original model!
+"""
+
+# Freeze the whole model.
+lora_model.trainable = False
```
Comment (mattdangerw) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:554]: why is this needed? i don't see any gradient descent in the rest of the guide

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 5]
```diff
@@ -0,0 +1,590 @@
...
+    query_lora_layer = (
+        lora_model.get_layer(gpt2_backbone_layer_name)
+        .get_layer(f"transformer_layer_{layer_idx}")
+        ._self_attention_layer._query_dense
+    )
+
+    A_weights = query_lora_layer.A.kernel  # (768, 1) (a, b)
+    B_weights = query_lora_layer.B.kernel  # (1, 12, 64) (b, c, d)
+    increment_weights = tf.einsum("ab,bcd->acd", A_weights, B_weights) * (ALPHA / RANK)
+    query_lora_layer.original_layer.kernel.assign_add(increment_weights)
```
Comment (mattdangerw) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:566]: very nice little snippet here!

**Review by mattdangerw — CHANGES_REQUESTED** [review]
Super impressive! Love this guide. Dropped a few comments.

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py (hunk 5 above)
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:566]: Thanks!

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py (hunk 4 above)
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:554]: Isn't it useful for demonstration purposes?

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+
+The original equation is `output = W0x + b0`, where `x` is the input, W0 and b0
+are the weight matrix and bias terms of the original dense layer (frozen). The
+LoRA equation is: `output = W0x + b0 + BAx`, where `A` and `B` are the
+rank-decomposition matrices.
+
+LoRA is based on the idea that updates to the weights of the pre-trained
+language model have a low "intrinsic rank" since pre-trained language models are
+over-parametrized. Predictive performance of full fine-tuning can be replicated
+even by constraining `W0`'s updates to low-rank decomposition matrices.
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:268]: Oh, yeah. I forgot to add this

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py (hunk 4 above)
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:554]: Isn't it useful for demonstration purposes? Can remove if it feel redundant

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 7]
```diff
@@ -0,0 +1,595 @@
...
+        trainable=True,
+    )
+
+"""
+Let's now do a forward pass to make sure we still have a valid chain of
+computation.
+"""
+
+lora_model(preprocessor(["they are going to ban LoRA in EU, lol"])[0])
+pass
```
Comment (fchollet) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:463]: Presumably this statement should be removed?

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 8]
```diff
@@ -0,0 +1,595 @@
...
+
+"""
+Freeze the entire LLM, only the LoRA layers should be trainable.
+"""
+
+for l in lora_model._flatten_layers():
```
Comment (fchollet) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:469]: Prefer using fully spelled out names everywhere, e.g. `layer` instead of `l`, for readability

**Review by fchollet — COMMENTED** [review]
Very nice work!

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+
+Now that we have hacked and verified the LoRA GPT-2 model, let's train it!
+"""
+
+gpu_memory_callback = GPUMemoryCallback(
+    target_batches=[5, 10, 25, 50, 100, 150, 200, 300, 400, 500],
+    print_stats=True,
+)
+
+optimizer = keras.optimizers.AdamW(
```
Comment (mattdangerw) [inline examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:503]: Can we just re-use the original optimizer? Or will that leak some memory from the first training run?

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py (hunk 4 above)
Comment (mattdangerw) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:564]: I guess to me I think of `trainable` as only pertaining to `fit` calls. So it threw me for a bit of a loop here and I was wondering if I missed something below. You could add a comment above the generate call maybe `# Freezing weights not necessary during generation, no weights are updated.`

**Review by mattdangerw — APPROVED** [review]
Some last comments, but LGTM from me!

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py (hunk 7 above)
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:467]: Oh, yeah. I'll pass some other string here

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+Last modified: 2023/05/27
+Description: Use KerasNLP to fine-tune a GPT-2 LLM with LoRA.
+Accelerator: GPU
+"""
+
+"""
+## Introduction
+
+Large Language Models (LLMs) have been shown to be effective at a variety of NLP
+tasks. An LLM is first pre-trained on a large corpus of text using general
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:14]: How about
```
An LLM is first pre-trained on a large corpus of text in a
self-supervised fashion.
```
?

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+    # Change query dense layer.
+    decoder_layer = lora_model.backbone.get_layer(f"transformer_layer_{layer_idx}")
+    self_attention_layer = decoder_layer._self_attention_layer
+
+    # Change query dense layer.
+    self_attention_layer._query_dense = LoraLayer(
+        self_attention_layer._query_dense,
+        rank=RANK,
+        alpha=ALPHA,
+        layer_idx=layer_idx,
```
[the remark this answers is not in the document]
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:448]: I'm going to call it "lora_A" and "lora_B", then

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py (hunk 9 above)
Comment (abheesht17) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py:506]: Pulled all of it in a function

File: examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply examples/nlp/parameter_efficient_finetuning_of_gpt2_with_lora.py]: sgtm!

**Comment by abheesht17** [comment]
@fchollet - yep! This PR is good to go

**Comment by fchollet** [comment]
Thank you for the great contribution! 👍
